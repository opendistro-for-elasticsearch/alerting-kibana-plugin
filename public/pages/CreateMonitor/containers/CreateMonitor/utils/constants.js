import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  searchType: 'graph',
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  aggregationType: 'count',
  fieldName: [],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  where: {
    fieldName: [],
    operator: OPERATORS_MAP.IS,
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  detectorId: '',
};
