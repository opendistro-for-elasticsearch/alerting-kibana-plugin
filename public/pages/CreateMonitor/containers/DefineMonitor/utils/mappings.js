/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export function shouldSkip(mapping) {
  const isDisabled = mapping.enabled === false;
  const hasIndexDisabled = mapping.index === false;
  const isNestedDataType = mapping.type === 'nested';
  return isDisabled || hasIndexDisabled || isNestedDataType;
}

export function resolvePath(path, field) {
  if (path) return `${path}.${field}`;
  return field;
}

export function getFieldsFromProperties(properties, dataTypes, path) {
  Object.entries(properties).forEach(([field, value]) => {
    getTypeFromMappings(value, dataTypes, resolvePath(path, field));
  });
}

export function getTypeFromMappings(mappings, dataTypes, path = '') {
  if (shouldSkip(mappings)) return dataTypes;
  // if there are properties then type is inherently an object
  if (mappings.properties) {
    getFieldsFromProperties(mappings.properties, dataTypes, path);
    return dataTypes;
  }

  const type = mappings.type;

  if (dataTypes[type]) dataTypes[type].add(path);
  else dataTypes[type] = new Set([path]);
  return dataTypes;
}

export function getPathsPerDataType(mappings) {
  const dataTypes = {};
  Object.entries(mappings)
    .forEach(([index, { mappings: docMappings }]) => getTypeFromMappings(docMappings, dataTypes));
  return dataTypes;
}
