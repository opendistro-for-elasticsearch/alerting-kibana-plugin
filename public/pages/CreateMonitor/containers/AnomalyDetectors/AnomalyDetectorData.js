/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { AppContext } from '../../../../utils/AppContext';
import { AD_PREVIEW_DAYS } from '../../../../utils/constants';

class AnomalyDetectorData extends React.Component {
  static contextType = AppContext;
  constructor(props) {
    super(props);
    this.state = {
      anomalyResult: {
        anomalies: [],
        featureData: {},
      },
      detector: {
        featureAttributes: [],
      },
      previewStartTime: 0,
      previewEndTime: 0,
      isLoading: false,
    };
    this.getPreviewData = this.getPreviewData.bind(this);
  }
  async componentDidMount() {
    await this.getPreviewData();
  }

  async getPreviewData() {
    const { detectorId, startTime, endTime } = this.props;
    const { httpClient } = this.context;
    this.setState({
      isLoading: true,
    });
    if (!detectorId) return;
    const requestParams = {
      startTime: moment()
        .subtract(AD_PREVIEW_DAYS, 'd')
        .valueOf(),
      startTime: startTime,
      endTime: endTime,
      preview: this.props.preview,
    };
    try {
      const response = await httpClient.get(`../api/alerting/detectors/${detectorId}/results`, {
        params: requestParams,
      });
      if (response.data.ok) {
        const { anomalyResult, detector } = response.data.response;
        this.setState({
          ...this.state,
          anomalyResult,
          detector,
          previewStartTime: requestParams.startTime,
          previewEndTime: requestParams.endTime,
          isLoading: false,
        });
      } else {
        this.setState({
          isLoading: false,
        });
      }
    } catch (err) {
      console.error('Unable to get detectorResults', err);
      this.setState({
        isLoading: false,
      });
    }
  }
  render() {
    const { render } = this.props;
    return render({ ...this.state });
  }
}

AnomalyDetectorData.propTypes = {
  detectorId: PropTypes.string.isRequired,
  preview: PropTypes.bool,
};
AnomalyDetectorData.defaultProps = {
  preview: true,
  startTime: moment()
    .subtract(5, 'd')
    .valueOf(),
  endTime: moment().valueOf(),
};

export { AnomalyDetectorData };
