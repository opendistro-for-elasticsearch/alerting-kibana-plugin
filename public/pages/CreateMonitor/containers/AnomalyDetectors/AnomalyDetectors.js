/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import { FormikComboBox } from '../../../../components/FormControls';
import { hasError, isInvalid, validateDetector } from '../../../../utils/validate';
import { AppContext } from '../../../../utils/AppContext';

class AnomalyDetectors extends React.Component {
  static contextType = AppContext;
  constructor(props) {
    super(props);
    this.state = {
      detectorOptions: [],
      isLoading: false,
    };
    this.searchDetectors = this.searchDetectors.bind(this);
  }
  async componentDidMount() {
    await this.searchDetectors();
  }

  async searchDetectors() {
    const { httpClient } = this.context;
    try {
      const response = await httpClient.post('../api/alerting/detectors/_search');
      if (response.data.ok) {
        const detectorOptions = response.data.detectors.map(detector => ({
          label: detector.name,
          value: detector.id,
          features: detector.featureAttributes,
          //TODO::Temp hack until backend fix
          interval: detector.detectionInterval,
        }));
        this.setState({ detectorOptions });
      }
    } catch (err) {
      console.error(err);
    }
  }

  render() {
    const { detectorOptions } = this.state;
    const { values, detectorId } = this.props;
    //Default to empty
    let selectedOptions = [];
    if (detectorOptions.length > 0) {
      const adId = values.detectorId ? values.detectorId : detectorId;
      const selectedValue = detectorOptions.find(detector => adId === detector.value);
      if (selectedValue) {
        selectedOptions = [selectedValue];
      }
    }
    return (
      <div>
        <FormikComboBox
          name={'detectorId'}
          formRow
          rowProps={{
            label: 'Detector',
            isInvalid,
            error: hasError,
          }}
          fieldProps={{
            validate: value => validateDetector(value, selectedOptions[0]),
          }}
          inputProps={{
            placeholder: 'Select a detector',
            options: detectorOptions,
            onBlur: (e, field, form) => {
              form.setFieldTouched('detectorId', true);
            },
            onChange: (options, field, form) => {
              form.setFieldError('detectorId', undefined);
              form.setFieldValue('detectorId', get(options, '0.value', ''));
              form.setFieldValue('period', {
                interval: 2 * get(options, '0.interval.period.interval'),
                unit: get(options, '0.interval.period.unit'),
              });
            },
            singleSelection: { asPlaintext: true },
            isClearable: false,
            selectedOptions,
          }}
        />
      </div>
    );
  }
}

AnomalyDetectors.propTypes = {
  values: PropTypes.object.isRequired,
  renderEmptyMessage: PropTypes.func.isRequired,
};

export default AnomalyDetectors;
