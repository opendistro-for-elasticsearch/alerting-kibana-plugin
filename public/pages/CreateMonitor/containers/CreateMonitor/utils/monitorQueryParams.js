import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = queryParams => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
  };
};
