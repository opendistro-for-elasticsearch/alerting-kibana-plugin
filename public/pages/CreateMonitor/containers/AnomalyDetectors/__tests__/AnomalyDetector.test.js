/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { Formik } from 'formik';
import { mount } from 'enzyme';

import { FORMIK_INITIAL_VALUES } from '../../CreateMonitor/utils/constants';
import AnomalyDetectors from '../AnomalyDetectors';
import { httpClientMock } from '../../../../../../test/mocks';
import { AppContext } from '../../../../../utils/AppContext';

jest.mock('ui/chrome', () => ({
  getBasePath: () => {
    return 'http://localhost/app';
  },
}));

const renderEmptyMessage = jest.fn();
function getMountWrapper() {
  return mount(
    <AppContext.Provider value={{ httpClient: httpClientMock }}>
      <Formik
        initialValues={FORMIK_INITIAL_VALUES}
        render={({ values }) => (
          <AnomalyDetectors values={values} renderEmptyMessage={renderEmptyMessage} />
        )}
      />
    </AppContext.Provider>
  );
}

describe('AnomalyDetectors', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders', () => {
    httpClientMock.post.mockResolvedValue({ data: { ok: true, detectors: [] } });
    const wrapper = getMountWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  test('should be able to select the detector', done => {
    httpClientMock.post.mockResolvedValueOnce({
      data: {
        ok: true,
        detectors: [{ name: 'sample-detector', id: 'sample-id', feature_attributes: [] }],
      },
    });
    //Preview mock
    httpClientMock.post.mockResolvedValueOnce({
      data: {
        ok: true,
        response: { anomalyResult: { anomalies: [], featureData: [] }, detector: {} },
      },
    });
    const wrapper = getMountWrapper();
    setTimeout(() => {
      wrapper
        .find('[data-test-subj="comboBoxSearchInput"]')
        .hostNodes()
        .simulate('change', { target: { value: 'sample-detector' } })
        .simulate('keyDown', { keyCode: 40 })
        .simulate('keyDown', { keyCode: 13 });
      expect(wrapper.instance().state.values.detectorId).toEqual('sample-id');
      done();
    });
  });
});
