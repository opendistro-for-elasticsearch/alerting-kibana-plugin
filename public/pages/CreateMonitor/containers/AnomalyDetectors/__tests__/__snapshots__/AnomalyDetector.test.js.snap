// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnomalyDetectors renders 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "aggregationType": "count",
      "bucketUnitOfTime": "h",
      "bucketValue": 1,
      "cronExpression": "0 */1 * * *",
      "daily": 0,
      "detectorId": "",
      "disabled": false,
      "fieldName": Array [],
      "frequency": "interval",
      "groupedOverFieldName": "bytes",
      "groupedOverTop": 5,
      "index": Array [],
      "monthly": Object {
        "day": 1,
        "type": "day",
      },
      "name": "",
      "overDocuments": "all documents",
      "period": Object {
        "interval": 1,
        "unit": "MINUTES",
      },
      "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
      "searchType": "graph",
      "timeField": "",
      "timezone": Array [],
      "weekly": Object {
        "fri": false,
        "mon": false,
        "sat": false,
        "sun": false,
        "thur": false,
        "tue": false,
        "wed": false,
      },
      "where": Object {
        "fieldName": Array [],
        "fieldRangeEnd": 0,
        "fieldRangeStart": 0,
        "fieldValue": "",
        "operator": "is",
      },
    }
  }
  isInitialValid={false}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
>
  <AnomalyDetectors
    renderEmptyMessage={[MockFunction]}
    values={
      Object {
        "aggregationType": "count",
        "bucketUnitOfTime": "h",
        "bucketValue": 1,
        "cronExpression": "0 */1 * * *",
        "daily": 0,
        "detectorId": "",
        "disabled": false,
        "fieldName": Array [],
        "frequency": "interval",
        "groupedOverFieldName": "bytes",
        "groupedOverTop": 5,
        "index": Array [],
        "monthly": Object {
          "day": 1,
          "type": "day",
        },
        "name": "",
        "overDocuments": "all documents",
        "period": Object {
          "interval": 1,
          "unit": "MINUTES",
        },
        "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
        "searchType": "graph",
        "timeField": "",
        "timezone": Array [],
        "weekly": Object {
          "fri": false,
          "mon": false,
          "sat": false,
          "sun": false,
          "thur": false,
          "tue": false,
          "wed": false,
        },
        "where": Object {
          "fieldName": Array [],
          "fieldRangeEnd": 0,
          "fieldRangeStart": 0,
          "fieldValue": "",
          "operator": "is",
        },
      }
    }
  >
    <div
      style={
        Object {
          "maxWidth": "390px",
        }
      }
    >
      <FormikComboBox
        fieldProps={
          Object {
            "validate": [Function],
          }
        }
        formRow={true}
        inputProps={
          Object {
            "isClearable": false,
            "onBlur": [Function],
            "onChange": [Function],
            "options": Array [],
            "placeholder": "Select a detector",
            "selectedOptions": Array [],
            "singleSelection": Object {
              "asPlaintext": true,
            },
          }
        }
        name="detectorId"
        rowProps={
          Object {
            "error": [Function],
            "isInvalid": [Function],
            "label": "Detector",
          }
        }
      >
        <FormikInputWrapper
          fieldProps={
            Object {
              "validate": [Function],
            }
          }
          name="detectorId"
          render={[Function]}
        >
          <C
            name="detectorId"
            render={[Function]}
            validate={[Function]}
          >
            <FieldInner
              formik={
                Object {
                  "dirty": false,
                  "errors": Object {},
                  "handleBlur": [Function],
                  "handleChange": [Function],
                  "handleReset": [Function],
                  "handleSubmit": [Function],
                  "initialValues": Object {
                    "aggregationType": "count",
                    "bucketUnitOfTime": "h",
                    "bucketValue": 1,
                    "cronExpression": "0 */1 * * *",
                    "daily": 0,
                    "detectorId": "",
                    "disabled": false,
                    "fieldName": Array [],
                    "frequency": "interval",
                    "groupedOverFieldName": "bytes",
                    "groupedOverTop": 5,
                    "index": Array [],
                    "monthly": Object {
                      "day": 1,
                      "type": "day",
                    },
                    "name": "",
                    "overDocuments": "all documents",
                    "period": Object {
                      "interval": 1,
                      "unit": "MINUTES",
                    },
                    "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
                    "searchType": "graph",
                    "timeField": "",
                    "timezone": Array [],
                    "weekly": Object {
                      "fri": false,
                      "mon": false,
                      "sat": false,
                      "sun": false,
                      "thur": false,
                      "tue": false,
                      "wed": false,
                    },
                    "where": Object {
                      "fieldName": Array [],
                      "fieldRangeEnd": 0,
                      "fieldRangeStart": 0,
                      "fieldValue": "",
                      "operator": "is",
                    },
                  },
                  "isSubmitting": false,
                  "isValid": false,
                  "isValidating": false,
                  "registerField": [Function],
                  "resetForm": [Function],
                  "setError": [Function],
                  "setErrors": [Function],
                  "setFieldError": [Function],
                  "setFieldTouched": [Function],
                  "setFieldValue": [Function],
                  "setFormikState": [Function],
                  "setStatus": [Function],
                  "setSubmitting": [Function],
                  "setTouched": [Function],
                  "setValues": [Function],
                  "submitCount": 0,
                  "submitForm": [Function],
                  "touched": Object {},
                  "unregisterField": [Function],
                  "validate": undefined,
                  "validateField": [Function],
                  "validateForm": [Function],
                  "validateOnBlur": true,
                  "validateOnChange": true,
                  "validationSchema": undefined,
                  "values": Object {
                    "aggregationType": "count",
                    "bucketUnitOfTime": "h",
                    "bucketValue": 1,
                    "cronExpression": "0 */1 * * *",
                    "daily": 0,
                    "detectorId": "",
                    "disabled": false,
                    "fieldName": Array [],
                    "frequency": "interval",
                    "groupedOverFieldName": "bytes",
                    "groupedOverTop": 5,
                    "index": Array [],
                    "monthly": Object {
                      "day": 1,
                      "type": "day",
                    },
                    "name": "",
                    "overDocuments": "all documents",
                    "period": Object {
                      "interval": 1,
                      "unit": "MINUTES",
                    },
                    "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
                    "searchType": "graph",
                    "timeField": "",
                    "timezone": Array [],
                    "weekly": Object {
                      "fri": false,
                      "mon": false,
                      "sat": false,
                      "sun": false,
                      "thur": false,
                      "tue": false,
                      "wed": false,
                    },
                    "where": Object {
                      "fieldName": Array [],
                      "fieldRangeEnd": 0,
                      "fieldRangeStart": 0,
                      "fieldValue": "",
                      "operator": "is",
                    },
                  },
                }
              }
              name="detectorId"
              render={[Function]}
              validate={[Function]}
            >
              <FormikFormRow
                form={
                  Object {
                    "dirty": false,
                    "errors": Object {},
                    "handleBlur": [Function],
                    "handleChange": [Function],
                    "handleReset": [Function],
                    "handleSubmit": [Function],
                    "initialValues": Object {
                      "aggregationType": "count",
                      "bucketUnitOfTime": "h",
                      "bucketValue": 1,
                      "cronExpression": "0 */1 * * *",
                      "daily": 0,
                      "detectorId": "",
                      "disabled": false,
                      "fieldName": Array [],
                      "frequency": "interval",
                      "groupedOverFieldName": "bytes",
                      "groupedOverTop": 5,
                      "index": Array [],
                      "monthly": Object {
                        "day": 1,
                        "type": "day",
                      },
                      "name": "",
                      "overDocuments": "all documents",
                      "period": Object {
                        "interval": 1,
                        "unit": "MINUTES",
                      },
                      "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
                      "searchType": "graph",
                      "timeField": "",
                      "timezone": Array [],
                      "weekly": Object {
                        "fri": false,
                        "mon": false,
                        "sat": false,
                        "sun": false,
                        "thur": false,
                        "tue": false,
                        "wed": false,
                      },
                      "where": Object {
                        "fieldName": Array [],
                        "fieldRangeEnd": 0,
                        "fieldRangeStart": 0,
                        "fieldValue": "",
                        "operator": "is",
                      },
                    },
                    "isSubmitting": false,
                    "isValid": false,
                    "isValidating": false,
                    "registerField": [Function],
                    "resetForm": [Function],
                    "setError": [Function],
                    "setErrors": [Function],
                    "setFieldError": [Function],
                    "setFieldTouched": [Function],
                    "setFieldValue": [Function],
                    "setFormikState": [Function],
                    "setStatus": [Function],
                    "setSubmitting": [Function],
                    "setTouched": [Function],
                    "setValues": [Function],
                    "submitCount": 0,
                    "submitForm": [Function],
                    "touched": Object {},
                    "unregisterField": [Function],
                    "validateField": [Function],
                    "validateForm": [Function],
                    "validateOnBlur": true,
                    "validateOnChange": true,
                    "values": Object {
                      "aggregationType": "count",
                      "bucketUnitOfTime": "h",
                      "bucketValue": 1,
                      "cronExpression": "0 */1 * * *",
                      "daily": 0,
                      "detectorId": "",
                      "disabled": false,
                      "fieldName": Array [],
                      "frequency": "interval",
                      "groupedOverFieldName": "bytes",
                      "groupedOverTop": 5,
                      "index": Array [],
                      "monthly": Object {
                        "day": 1,
                        "type": "day",
                      },
                      "name": "",
                      "overDocuments": "all documents",
                      "period": Object {
                        "interval": 1,
                        "unit": "MINUTES",
                      },
                      "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
                      "searchType": "graph",
                      "timeField": "",
                      "timezone": Array [],
                      "weekly": Object {
                        "fri": false,
                        "mon": false,
                        "sat": false,
                        "sun": false,
                        "thur": false,
                        "tue": false,
                        "wed": false,
                      },
                      "where": Object {
                        "fieldName": Array [],
                        "fieldRangeEnd": 0,
                        "fieldRangeStart": 0,
                        "fieldValue": "",
                        "operator": "is",
                      },
                    },
                  }
                }
                name="detectorId"
                rowProps={
                  Object {
                    "error": [Function],
                    "isInvalid": [Function],
                    "label": "Detector",
                  }
                }
              >
                <EuiFormRow
                  describedByIds={Array []}
                  fullWidth={false}
                  hasEmptyLabelSpace={false}
                  id="detectorId"
                  isInvalid={false}
                  label="Detector"
                  labelType="label"
                >
                  <div
                    className="euiFormRow"
                    id="detectorId-row"
                  >
                    <div>
                      <EuiFormLabel
                        aria-invalid={false}
                        htmlFor="detectorId"
                        isFocused={false}
                        isInvalid={false}
                        type="label"
                      >
                        <label
                          aria-invalid={false}
                          className="euiFormLabel"
                          htmlFor="detectorId"
                        >
                          Detector
                        </label>
                      </EuiFormLabel>
                    </div>
                    <ComboBox
                      field={
                        Object {
                          "name": "detectorId",
                          "onBlur": [Function],
                          "onChange": [Function],
                          "value": "",
                        }
                      }
                      form={
                        Object {
                          "dirty": false,
                          "errors": Object {},
                          "handleBlur": [Function],
                          "handleChange": [Function],
                          "handleReset": [Function],
                          "handleSubmit": [Function],
                          "initialValues": Object {
                            "aggregationType": "count",
                            "bucketUnitOfTime": "h",
                            "bucketValue": 1,
                            "cronExpression": "0 */1 * * *",
                            "daily": 0,
                            "detectorId": "",
                            "disabled": false,
                            "fieldName": Array [],
                            "frequency": "interval",
                            "groupedOverFieldName": "bytes",
                            "groupedOverTop": 5,
                            "index": Array [],
                            "monthly": Object {
                              "day": 1,
                              "type": "day",
                            },
                            "name": "",
                            "overDocuments": "all documents",
                            "period": Object {
                              "interval": 1,
                              "unit": "MINUTES",
                            },
                            "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
                            "searchType": "graph",
                            "timeField": "",
                            "timezone": Array [],
                            "weekly": Object {
                              "fri": false,
                              "mon": false,
                              "sat": false,
                              "sun": false,
                              "thur": false,
                              "tue": false,
                              "wed": false,
                            },
                            "where": Object {
                              "fieldName": Array [],
                              "fieldRangeEnd": 0,
                              "fieldRangeStart": 0,
                              "fieldValue": "",
                              "operator": "is",
                            },
                          },
                          "isSubmitting": false,
                          "isValid": false,
                          "isValidating": false,
                          "registerField": [Function],
                          "resetForm": [Function],
                          "setError": [Function],
                          "setErrors": [Function],
                          "setFieldError": [Function],
                          "setFieldTouched": [Function],
                          "setFieldValue": [Function],
                          "setFormikState": [Function],
                          "setStatus": [Function],
                          "setSubmitting": [Function],
                          "setTouched": [Function],
                          "setValues": [Function],
                          "submitCount": 0,
                          "submitForm": [Function],
                          "touched": Object {},
                          "unregisterField": [Function],
                          "validateField": [Function],
                          "validateForm": [Function],
                          "validateOnBlur": true,
                          "validateOnChange": true,
                          "values": Object {
                            "aggregationType": "count",
                            "bucketUnitOfTime": "h",
                            "bucketValue": 1,
                            "cronExpression": "0 */1 * * *",
                            "daily": 0,
                            "detectorId": "",
                            "disabled": false,
                            "fieldName": Array [],
                            "frequency": "interval",
                            "groupedOverFieldName": "bytes",
                            "groupedOverTop": 5,
                            "index": Array [],
                            "monthly": Object {
                              "day": 1,
                              "type": "day",
                            },
                            "name": "",
                            "overDocuments": "all documents",
                            "period": Object {
                              "interval": 1,
                              "unit": "MINUTES",
                            },
                            "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
                            "searchType": "graph",
                            "timeField": "",
                            "timezone": Array [],
                            "weekly": Object {
                              "fri": false,
                              "mon": false,
                              "sat": false,
                              "sun": false,
                              "thur": false,
                              "tue": false,
                              "wed": false,
                            },
                            "where": Object {
                              "fieldName": Array [],
                              "fieldRangeEnd": 0,
                              "fieldRangeStart": 0,
                              "fieldValue": "",
                              "operator": "is",
                            },
                          },
                        }
                      }
                      id="detectorId"
                      inputProps={
                        Object {
                          "isClearable": false,
                          "onBlur": [Function],
                          "onChange": [Function],
                          "options": Array [],
                          "placeholder": "Select a detector",
                          "selectedOptions": Array [],
                          "singleSelection": Object {
                            "asPlaintext": true,
                          },
                        }
                      }
                      name="detectorId"
                      onBlur={[Function]}
                      onFocus={[Function]}
                    >
                      <EuiComboBox
                        compressed={false}
                        fullWidth={false}
                        id="detectorId"
                        isClearable={false}
                        name="detectorId"
                        onBlur={[Function]}
                        onChange={[Function]}
                        options={Array []}
                        placeholder="Select a detector"
                        selectedOptions={Array []}
                        singleSelection={
                          Object {
                            "asPlaintext": true,
                          }
                        }
                      >
                        <div
                          aria-expanded={false}
                          aria-haspopup="listbox"
                          className="euiComboBox"
                          name="detectorId"
                          onKeyDown={[Function]}
                          role="combobox"
                        >
                          <EuiComboBoxInput
                            autoSizeInputRef={[Function]}
                            compressed={false}
                            focusedOptionId={null}
                            fullWidth={false}
                            hasSelectedOptions={false}
                            id="detectorId"
                            inputRef={[Function]}
                            isListOpen={false}
                            noIcon={false}
                            onChange={[Function]}
                            onClick={[Function]}
                            onCloseListClick={[Function]}
                            onFocus={[Function]}
                            onOpenListClick={[Function]}
                            onRemoveOption={[Function]}
                            placeholder="Select a detector"
                            rootId={[Function]}
                            searchValue=""
                            selectedOptions={Array []}
                            singleSelection={
                              Object {
                                "asPlaintext": true,
                              }
                            }
                            toggleButtonRef={[Function]}
                            updatePosition={[Function]}
                            value=""
                          >
                            <EuiFormControlLayout
                              compressed={false}
                              fullWidth={false}
                              icon={
                                Object {
                                  "aria-label": "Open list of options",
                                  "data-test-subj": "comboBoxToggleListButton",
                                  "disabled": undefined,
                                  "onClick": [Function],
                                  "ref": [Function],
                                  "side": "right",
                                  "type": "arrowDown",
                                }
                              }
                            >
                              <div
                                className="euiFormControlLayout"
                              >
                                <div
                                  className="euiFormControlLayout__childrenWrapper"
                                >
                                  <div
                                    className="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap"
                                    data-test-subj="comboBoxInput"
                                    onClick={[Function]}
                                    tabIndex="-1"
                                  >
                                    <p
                                      className="euiComboBoxPlaceholder"
                                    >
                                      Select a detector
                                    </p>
                                    <AutosizeInput
                                      aria-activedescendant={null}
                                      aria-controls={null}
                                      className="euiComboBox__input"
                                      data-test-subj="comboBoxSearchInput"
                                      id="detectorId"
                                      injectStyles={true}
                                      inputRef={[Function]}
                                      minWidth={1}
                                      onBlur={[Function]}
                                      onChange={[Function]}
                                      onFocus={[Function]}
                                      role="textbox"
                                      style={
                                        Object {
                                          "fontSize": 14,
                                        }
                                      }
                                      value=""
                                    >
                                      <div
                                        className="euiComboBox__input"
                                        style={
                                          Object {
                                            "display": "inline-block",
                                            "fontSize": 14,
                                          }
                                        }
                                      >
                                        <input
                                          aria-activedescendant={null}
                                          aria-controls={null}
                                          data-test-subj="comboBoxSearchInput"
                                          id="detectorId"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          role="textbox"
                                          style={
                                            Object {
                                              "boxSizing": "content-box",
                                              "width": "2px",
                                            }
                                          }
                                          value=""
                                        />
                                        <div
                                          style={
                                            Object {
                                              "height": 0,
                                              "left": 0,
                                              "overflow": "scroll",
                                              "position": "absolute",
                                              "top": 0,
                                              "visibility": "hidden",
                                              "whiteSpace": "pre",
                                            }
                                          }
                                        />
                                      </div>
                                    </AutosizeInput>
                                  </div>
                                  <EuiFormControlLayoutIcons
                                    icon={
                                      Object {
                                        "aria-label": "Open list of options",
                                        "data-test-subj": "comboBoxToggleListButton",
                                        "disabled": undefined,
                                        "onClick": [Function],
                                        "ref": [Function],
                                        "side": "right",
                                        "type": "arrowDown",
                                      }
                                    }
                                  >
                                    <div
                                      className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                                    >
                                      <EuiFormControlLayoutCustomIcon
                                        aria-label="Open list of options"
                                        data-test-subj="comboBoxToggleListButton"
                                        iconRef={[Function]}
                                        onClick={[Function]}
                                        type="arrowDown"
                                      >
                                        <button
                                          aria-label="Open list of options"
                                          className="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                                          data-test-subj="comboBoxToggleListButton"
                                          onClick={[Function]}
                                          type="button"
                                        >
                                          <EuiIcon
                                            aria-hidden="true"
                                            className="euiFormControlLayoutCustomIcon__icon"
                                            type="arrowDown"
                                          >
                                            <div>
                                              EuiIconMock
                                            </div>
                                          </EuiIcon>
                                        </button>
                                      </EuiFormControlLayoutCustomIcon>
                                    </div>
                                  </EuiFormControlLayoutIcons>
                                </div>
                              </div>
                            </EuiFormControlLayout>
                          </EuiComboBoxInput>
                        </div>
                      </EuiComboBox>
                    </ComboBox>
                  </div>
                </EuiFormRow>
              </FormikFormRow>
            </FieldInner>
          </C>
        </FormikInputWrapper>
      </FormikComboBox>
    </div>
  </AnomalyDetectors>
</Formik>
`;
